### PR TITLE
Add table properties for number of entries added to filters

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: With Ribbon filter work and possible variance in actual bits
per key (or prefix; general term "entry") to achieve certain FP rates,
I've received a request to be able to track actual bits per key in
generated filters. This change adds a num_filter_entries table
property, which can be combined with filter_size to get bits per key
(entry).

This can vary from num_entries in at least these ways:
* Different versions of same key are only counted once in filters.
* With prefix filters, several user keys map to the same filter entry.
* A single filter can include both prefixes and user keys.

Note that FilterBlockBuilder::NumAdded() didn't do anything useful
except distinguish empty from non-empty.

Test Plan: basic unit test included, others updated

### CODE REVIEW

File: table/block_based/partitioned_filter_block.cc [hunk 1]
```diff
@@ -131,6 +134,9 @@ Slice PartitionedFilterBlockBuilder::Finish(
   if (UNLIKELY(filters.empty())) {
     *status = Status::OK();
     if (finishing_filters) {
+      // Simplest to just add them all at the end
+      *num_entries_added += total_added_in_built_;
```
Comment (jay-zhuang) [inline table/block_based/partitioned_filter_block.cc:138]: nit: check `num_entries_added` not nullptr?

**Review by jay-zhuang — APPROVED** [review]
LGTM.

There was discussion about memory usage of table properties, I'm wondering how much memory will this change add?

File: table/block_based/full_filter_block.h [hunk 2]
```diff
@@ -79,7 +80,7 @@ class FullFilterBlockBuilder : public FilterBlockBuilder {
   // last_key_in_domain_ is true, regardless of the current key.
   bool last_key_in_domain_;
 
-  uint32_t num_added_;
+  uint32_t any_added_;
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply table/block_based/full_filter_block.h:83]: oops!

File: table/block_based/full_filter_block.cc [hunk 3]
```diff
@@ -97,13 +97,14 @@ void FullFilterBlockBuilder::Reset() {
...
+Slice FullFilterBlockBuilder::Finish(const BlockHandle& /*tmp*/, Status* status,
+                                     uint64_t* num_entries_added) {
   Reset();
   // In this impl we ignore BlockHandle
   *status = Status::OK();
-  if (num_added_ != 0) {
-    num_added_ = 0;
+  if (any_added_) {
+    any_added_ = false;
+    *num_entries_added += filter_bits_builder_->GetNumEntriesAdded();
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply table/block_based/full_filter_block.cc:107]: I think what happened is that I was expecting to have to accumulate between calls to Finish() for partitioned filters, and then found I needed to accumulate internally in PartitionedFilterBlockBuilder before any Finish() and just return the final sum at the end. And I forgot that I could go back and simplify the interface. Thanks.

**Comment by pdillinger** [comment]
> There was discussion about memory usage of table properties, I'm wondering how much memory will this change add?

Should be eight bytes per SST file. I'm more concerned about string properties and how they currently involve indirection (multiple objects, std::string and its contents).

File: table/block_based/full_filter_block.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (pdillinger) [reply table/block_based/full_filter_block.cc:107]: A complication is that an old custom FilterBitsBuilder won't report an accurate number, just 0. So we can't trust 0 to mean "don't generate filter." We need some independent tracking of "was anything added?"

**Pull request closed without merging**